### DISCUSSION TITLE
Does Slint implement a file picker?

### DISCUSSION THREAD

**Discussion by adrian5** [slint-ui/slint#1959] [Q&A] [answered]
I haven't given Slint a try yet, but I was wondering about the degree of Desktop OS integration. Does Slint supply its own file picker dialog, hook into the one from the OS, or how is this handled?

**Comment by ogoffart** [comment, marked as answer]
Slint currently doesn't provide its own file picker, but relies on existing ones.  
For example, in Rust, you can use the [rfd crate](https://lib.rs/crates/rfd).
(An example of that is [cargo-ui](https://github.com/slint-ui/cargo-ui)  which has a open file dialog when you press the "📂" button)

**Reply by bowenxuuu** [reply]
Hello @ogoffart 
How can I open the `file-dialog` in a modal state?

```
let file = FileDialog::new()
            .set_parent(xxxx)
            .add_filter("Excel", &["xls", "xlsx"])
            .set_directory("/")
            .pick_file()
            .unwrap_or_default();
```

The FileDialog needs a parent window to open in a modal state, but I don't know how to get a parent window in Slint.

**Reply by ogoffart** [reply]
You can use [Window::window_handle](https://docs.rs/slint/latest/slint/struct.Window.html#method.window_handle) to get a raw window handle which can be used in set_parent

**Reply by bowenxuuu** [reply]
Thanks.

**Comment by ultimaweapon** [comment]
We have used `rfd` in the past and found out it block Slint event loop since it is never return until the user exit the browsing dialog. On Windows this probably fine but on Linux with Wayland the compositor will think your application is unresponsive and show the popup asking the user to wait or kill your application. We end up implementing our own file picker dialog to fix this issue.

**Reply by ogoffart** [reply]
You can open the FileDialog from a thread.
Or have you tried `rfd::AsyncFileDialog` and running that with `slint::spawn_local` ?

**Reply by ultimaweapon** [reply]
Nope we have not tried both of it. After evaluating `rfd` source I believe it is better for us to implement our own one. Can't remember exact reason behind this but we are happy with our own implementation.

**Reply by over2sd** [reply]
Is your implementation of file picker code available anywhere?

**Reply by ultimaweapon** [reply]
Yes but it is rely on our custom made [Platform](https://docs.rs/slint/latest/slint/platform/trait.Platform.html):

- Linux: https://github.com/obhq/obliteration/blob/main/gui/src/ui/linux/dialogs.rs
- macOS: https://github.com/obhq/obliteration/blob/main/gui/src/ui/macos/dialogs.rs
- Windows: https://github.com/obhq/obliteration/blob/main/gui/src/ui/windows/dialogs.rs

All above code dual licensed under MIT License and Apache License, Version 2.0.

**Comment by asmaier** [comment]
Is a file picker available for other languages, e.g, python?

**Reply by ogoffart** [reply]
Not at the moment. This is traced at https://github.com/slint-ui/slint/issues/9781 , please add a :+1: there.
